### DISCUSSION TITLE
Accessing internal integrator time step and state values (IDAS/CVODES)

### DISCUSSION THREAD

**Discussion by djlaky** [casadi/casadi#3757] [Q&A]
Hello!

I was wondering if there is a way to access the time steps that the integrator has chosen with variable step size from the integrate() function call. For reference, I was hoping to get information in the following link from the SUNDIALS documentation, the option of ``ONE_STEP`` for the ``CVodeSetStopTime`` field produces the time points the integrator chose (not the ones provided _a priori_ to the integrator with the ``grid`` option within the python CasADi interface).

[https://computing.llnl.gov/projects/sundials/usage-notes](https://computing.llnl.gov/projects/sundials/usage-notes)

For context, I am interested in using the time steps to build a simultaneous, equation-oriented model for optimization using an AML (Pyomo in this case) where the internal time steps may be important to capture the stiffness of the system. For instance, I have a model where in a certain region of the time domain I should use fine time steps and in another region of the time domain I should use coarse time steps. However, I have no way to know where these regions in the time domain would be beforehand, thus leveraging internal time steps taken by CVODES/IDAS would help me build a simultaneous, equation-oriented model that could possibly capture the stiffness of the system appropriately.

I did not notice any options I can pass to the integrator that will allow my output to have this format. I also did some searching and I couldn't find any topics that specifically addressed this. If you could help me out or point me to examples/resources that could work for this task, that would be great!

**Comment by atharvkumawat97** [comment]
Did you get the solution for this problem?

**Reply by djlaky** [reply]
I have not tried to implement this yet, but it may be possible with the ``monitor`` keyword. https://github.com/Pyomo/pyomo/issues/3718

**Reply by jaeandersson** [reply]
`monitor` should probably work
